guppylang/prelude/_internal/util.py

```diff
@@ -0,0 +1,417 @@
+) -> ht.Type:
+    """Makes a concrete hugr type using a guppy instantiation.
+
+    Args:
+        ty: The hugr type to make concrete.
+        inst: The guppy instantiation of the type arguments.
+        variable_remap: A mapping from the hugr param variable indices to
+            de Bruijn indices in the guppy type. Defaults to identity.
+    """
+    remap = variable_remap or {}
+
+    if isinstance(ty, ht.Variable) and remap.get(ty.idx, ty.idx) < len(inst):
+        concrete_arg: Argument = inst[remap.get(ty.idx, ty.idx)]
+        assert isinstance(
+            concrete_arg, TypeArg
+        ), f"Cannot translate const type {concrete_arg} arg into a type"
+        return concrete_arg.ty.to_hugr()
+    elif isinstance(ty, ht.Opaque):
+        # TODO: This is a temporary hack to compute bounds that depend on the
+        # type parameters. This won't be needed once we start using hugr's
+        # extension definitions, which include a `TypeDefBound` field to compute
+        # bounds during instantiation.
+        if ty.id == "List":
+            assert isinstance(inst[0], TypeArg)
+            bound = inst[0].ty.hugr_bound
+        else:
+            bound = ty.bound
+
+        return ht.Opaque(
+            id=ty.id,
+            bound=bound,
+            extension=ty.extension,
+            args=[make_concrete_arg(arg, inst, remap) for arg in ty.args],
+        )
+    elif isinstance(ty, ht.ExtType):
+        return ht.ExtType(
+            type_def=ty.type_def,
+            args=[make_concrete_arg(arg, inst, remap) for arg in ty.args],
+        )
+    return ty
+
+
+def custom_op(
+    name: str,
+    inp: Sequence[guppylang.tys.ty.Type | builtins.type | ht.Type],
+    out: Sequence[guppylang.tys.ty.Type | builtins.type | ht.Type],
+    args: list[ht.TypeArg],
+    *,
+    ext: str = "guppy.unsupported",
+    variable_remap: dict[int, int] | None = None,
+) -> Callable[[Inst], ops.DataflowOp]:
+    """Custom hugr operation
+
+    Args:
+        op_name: The name of the operation.
+        inp: The input types of the operation.
+            If the types are guppylang types, they are translated to Hugr types.
+            Any other types is treated as a native python type, and translated to
+            a Hugr type using `builtin_type_to_hugr`.
+        out: The output types of the operation.
+            If the types are guppylang types, they are translated to Hugr types.
+            Any other types is treated as a native python type, and translated to
+            a Hugr type using `builtin_type_to_hugr`.
+        args: The type arguments of the operation.
+        ext: The extension of the operation. Defaults to a placeholder extension.
+        variable_remap: A mapping from the hugr param variable indices to
+            de Bruijn indices in the guppy type. Defaults to identity.
+
+    Returns:
+        A function that takes an instantiation of the type arguments and returns
+        a concrete HUGR op.
```

Comment:
I think it would be better to take the whole (monomorphic) signature instead of the type args. `CustomCallCompiler.compile_call` can should be able to provide this signature.

Then you could get rid of most of the hacky things in this file